coordinator/src/funding_fee.rs

```diff
@@ -0,0 +1,277 @@
+    pub amount: SignedAmount,
+    pub trader_pubkey: PublicKey,
+    pub position_id: i32,
+    pub due_date: OffsetDateTime,
+    pub paid_date: Option<OffsetDateTime>,
+}
+
+#[derive(Clone, Copy, Debug, serde::Serialize, serde::Deserialize, PartialEq)]
+pub enum IndexPriceSource {
+    Bitmex,
+    Fake,
+}
+
+pub async fn generate_funding_fee_events_periodically(
+    scheduler: &JobScheduler,
+    pool: Pool<ConnectionManager<PgConnection>>,
+    schedule: String,
+    index_price_source: IndexPriceSource,
+) -> Result<()> {
+    scheduler
+        .add(tokio_cron_scheduler::Job::new(
+            schedule.as_str(),
+            move |_, _| {
+                if let Err(e) = generate_funding_fee_events(&pool, index_price_source) {
+                    tracing::error!("Failed to generate funding fee events: {e:#}");
+                }
+            },
+        )?)
+        .await?;
+
+    scheduler.start().await?;
+
+    Ok(())
+}
+
+/// Generate [`FundingFeeEvent`]s for all active positions.
+///
+/// When called, a [`FundingFeeEvent`] will be generated for an active position if:
+///
+/// - We can get a [`FundingRate`] that is at most 1 hour old from the DB.
+/// - We can get a BitMEX index price for the `end_date` of the [`FundingRate`].
+/// - There is no other [`FundingFeeEvent`] in the DB with the same `position_id` and `end_date`.
+/// - The position was created _before_ the `end_date` of the [`FundingRate`].
+fn generate_funding_fee_events(
+    pool: &Pool<ConnectionManager<PgConnection>>,
+    index_price_source: IndexPriceSource,
+) -> Result<()> {
+    let mut conn = pool.get()?;
+
+    let funding_rate = db::funding_rates::get_funding_rate_charged_in_the_last_hour(&mut conn)?;
+
+    let funding_rate = match funding_rate {
+        Some(funding_rate) => funding_rate,
+        None => {
+            tracing::debug!("No current funding rate for this hour");
+            return Ok(());
+        }
+    };
+
+    // TODO: Funding rates should be specific to contract symbols.
+    let contract_symbol = ContractSymbol::BtcUsd;
+
+    let current_index_price = match index_price_source {
+        IndexPriceSource::Bitmex => block_in_place(move || {
+            let current_index_price =
+                get_bitmex_index_price(&contract_symbol, funding_rate.end_date)?;
+
+            anyhow::Ok(current_index_price)
+        })?,
+        IndexPriceSource::Fake => {
+            #[cfg(not(debug_assertions))]
+            compile_error!("Cannot use a fake index price in release mode");
+
+            dec!(50_000)
+        }
+    };
+
+    if current_index_price.is_zero() {
+        bail!("Cannot generate funding fee events with zero index price");
+    }
+
+    // We exclude active positions which were open after this funding period ended.
+    let positions = db::positions::Position::get_all_active_positions_open_before(
+        &mut conn,
+        funding_rate.end_date,
+    )?;
+    for position in positions {
+        let amount =
```

Comment:
Hmm, `calculate_funding_fee` already returns a `SignedAmount`. If the value is positive, the trader pays. If it's negative, the coordinator pays.

Reply:
but that's not the whole story, because the returned `SignedAmount` will depend on whether `funding_rate` is positive or negative. But whether the user has to pay or not depends on the position's direction.

Reply:
You are right!